### ISSUE TITLE
Theme namespace

### ISSUE THREAD

**Issue by lbelinsk** [wix/stylable#411]
I have this use case where I want to apply theme namespace:

```
:import {
  -st-from: "../theme.st.css";
  -st-default: BOTheme;
}
```

![image](https://user-images.githubusercontent.com/18248989/44086713-17d87b26-9fc6-11e8-80e7-f674cce9854f.png)

I am Getting 
`[stylable] unscoped native element "BOTheme" will affect all elements of the same type in the document`. Other than that warning, it will be great to have some nice syntax to apply only once at the top of the stylesheet that I am "root scoped" to this theme, instead of repeating this over and over for each rule in the stylesheet.

**Comment by lbelinsk** [comment]
Another issue that might get resolved automatically by fixing this:

// button.st.css
```
BOTheme .premium {
  background-color: #AA4DC8;
  border-color: #AA4DC8;
  color: white;
}

```

// dropdown.st.css
```
:import {
  -st-from: "../button/button.st.css";
  -st-named: premium;
}

BOTheme Dropdown::actionButton {
  -st-mixin: premium; // Not working since premium class is namespaced under the theme in button.st.css
}
```

**Comment by tomrav** [comment]
I'm starting work on this feature, we've come up with the syntax below.
It is limited to a single level (no inner nesting), but will allow for multiple different scopes in the same file. By using `@st-scope` the warning you've been getting will be ignored and mixins should work.

Appreciate any comments you might have.

```css
:import {
    -st-from: "some-package/BOTheme.st.css";
    -st-default: BOTheme;
}

@st-scope BOTheme {
    .root{}
    .btn {}
};
```

**Comment by mykas** [comment]
Looks pretty good, Tom.  @hepiyellow  what do you think?

**Comment by lbelinsk** [comment]
@tomrav just to make sure I understand, all of the rules inside 
```
@st-scope BOTheme { ... }
```
Will be scoped, just like we do today:
```
BOTheme  .someClass {}
BOTheme  .someClass2 {}
```

If so, sounds great

And what do you mean by `no inner nesting`? Can you give an example of an unsupported syntax?

**Comment by lbelinsk** [comment]
Also, what about supporting this:
```
BackofficeTheme .button,
BackofficeTheme.button {
    .... rules....
}
```

Placing the theme on the component itself, not on some wrapper

**Comment by mykas** [comment]
About inner nesting. Maybe he means that no such cases:
```css
@st-scope BOTheme { 
    @st-scope BOTheme { ... }
 }

**Comment by tomrav** [comment]
> ```
> @st-scope BOTheme { ... }
> ```
> Will be scoped, just like we do today:
> 
> ```
> BOTheme  .someClass {}
> BOTheme  .someClass2 {}
> ```

That's correct.


> And what do you mean by `no inner nesting`? Can you give an example of an unsupported syntax?

Exactly what @mykas has replied, it would not support defining an `@st-scope` inside an existing `@st-scope`, though you could create several (sibling) scopes in the same file.

---
> ```
> BackofficeTheme .button,
> BackofficeTheme.button {
>    .... rules....
> }
> ```

@lbelinsk I'm not exactly sure what's the usecase for this example, at the moment `@st-scope`will not support that capability.

**Comment by lbelinsk** [comment]
The usecase for us now is that we gradually need to start migrating components to support the new pattern of single wrapping root for the theme.

The goal is to support the following syntax:
```
<div className={backoffice}>
   <Button>Hey</Button>
</div>
```

But today we would like to change the `Button` implementation to have the `className={backoffice}` on its root to prevent a breaking change on the user's apps, so the Button will work for them even if it is not yet wrapped in the backoffice theme.
For that case we need the rule `BackofficeTheme.button`, not `BackofficeTheme .button`

**Comment by hepiyellow** [comment]
Hi,
The first issue about the `unscoped native element` was not discussed here.
So I opened an issue specifically for that.
#450

**Comment by tomrav** [comment]
> Hi,
> The first issue about the `unscoped native element` was not discussed here.
> So I opened an issue specifically for that.
> #450

By using the `@st-scope` command this specific error will not be generated.

**Comment by tomrav** [comment]
> Hey @tomrav, when can we expect `-st-scope` in stylable?

Probably sometime around the middle of next week.

**Comment by tomrav** [comment]
Resolved in #455
The error message has been made more clear, and `@st-scope` feature introduced to allow for controlled scoping.

**Comment by mykas** [comment]
Thanks Tom!
